Feature/data layer integration

mmathis21 opened the pull request:
Integrated data layer with API calls. Assets and facilities are now fetched from respective endpoints. Mock repositories are disconnected but left some if the code is needed in the future.

mmathis21 committed:
Added static string for API connection. Change localhost to 10.0.2.2 running on android

mmathis21 committed:
Removed point conversion, added fromJson and toJson metods. Changed all methods to future type.

mmathis21 committed:
Removed point class with file

mmathis21 committed:
Updated and cleaned up pubspec.yaml

mmathis21 committed:
Implemented fetchAssets(), returns all assets from /assets endpoint

mmathis21 committed:
Implemented getAll method to fetch and parse assets from the remote data source, mapped raw JSON data to a list of 'Asset' objects. Added error handling

mmathis21 committed:
Made an asynchronous getAll method to the repository interface.

mmathis21 committed:
Refactored methods to make them asynchronous

mmathis21 committed:
Removed mock repositories and connected the services to real repositories.

mmathis21 committed:
Refactored 'Point' to 'x' and 'y' for easier usage.

mmathis21 committed:
Assets page now uses a future builder, 'AssetList' displays a list of asset items.
Included loading and error handling.
`_AssetWithFacilityName` helper class pairs assets with resolved facility.

mmathis21 committed:
Refactored parentFacilityId to floorMapId, in accordance with JSON field name.

mmathis21 committed:
Added FacilityRemoteDataSource for fetching facilities from API

mmathis21 committed:
Implemented FacilityRepository with API integration for facilities and error handling.

mmathis21 committed:
Refactored Repository interface to include getAll() method.

mmathis21 committed:
Refactored methods to asynchronous.

mmathis21 committed:
Refactored facility entity to include toJson and fromJson methods.

mmathis21 committed:
Added base64String decoding. Images now appear in the facility card.

mmathis21 committed:
Removed unused import

mmathis21 committed:
Refactored methods to Future type.

mmathis21 committed:
Refactored 'Point' fields to 'x' and 'y'
Refactored methods to be asynchronous.

mmathis21 committed:
Refactored methods to be asynchronous.

mmathis21 committed:
Refactored methods to be asynchronous.
*Mock repository is disconnected*

mmathis21 committed:
Removed mock facility repository.

mmathis21 committed:
Refactored 'Point' fields to 'x' and 'y'.
Refactored parentFacilityId to floorMapId.

mmathis21 committed:
Disconnected zone history log repository. May be used later.

mmathis21 committed:
Refactored methods to be asynchronous.
*Mock repository is not in use.*

mmathis21 committed:
Disconnected mock asset position history log repository. May be used later.

mmathis21 committed:
Added FutureBuilders to load assets and facilities. FutureBuilders handle asynchronous loading.
Loading indicators and error messages for both data fetching processes

vdraganic21 commented on Android/indoor_localization/lib/data/connections/api.dart:
Fetch the API base url from config file.

mmathis21 replied:
Which config file?

vdraganic21 replied:
Resolved on daily meeting.

vdraganic21 commented on Android/indoor_localization/lib/data/datasources/remote/facility_remote_datasource.dart:
```diff
@@ -0,0 +1,23 @@
+
+  FacilityRemoteDataSource({required this.client});
+
+  // Fetch facilities from the API
```
Remove comment.

vdraganic21 commented on Android/indoor_localization/lib/data/repositories/facility_repository.dart:
```diff
@@ -0,0 +1,45 @@
+  }
+
+  @override
+  Future<Facility?> get(int id) async {
```
There is a API endpoint that returns a facility based on id. Use that instead of fetching all facilities and sorting out the one with the wanted id.

vdraganic21 commented on Android/indoor_localization/lib/domain/entities/asset.dart:
```diff
@@ -1,66 +1,101 @@
+
+  Future<int> getPosition() async {
+    final updatedAsset = await updateData();
+    return updatedAsset.x;
```
Only returns x for position?

vdraganic21 commented on Android/indoor_localization/lib/domain/mock-repositories/mock_asset_zone_history_log_repository.dart:
```diff
@@ -1,7 +1,7 @@
 import 'package:indoor_localization/domain/entities/asset_zone_history_log_filter.dart';
 import 'package:indoor_localization/domain/repository-interfaces/asset_zone_history_log_repository.dart';
-
+/*
```
Why comment out the mock classes?

mmathis21 replied:
Not used right now, in case someone needs it later.

vdraganic21 commented on Android/indoor_localization/lib/domain/mock-repositories/mock_facility_repository.dart:
Don't delete the mock classes.

vdraganic21 commented on Android/indoor_localization/lib/domain/repository-interfaces/asset_repository.dart:
Valid, these methods should be async.

vdraganic21 commented on Android/indoor_localization/lib/domain/entities/asset.dart:
```diff
@@ -1,66 +1,101 @@
-  Facility? getCurrentFacility() {
-    final facility = FacilityService.get(parentFacilityId);
-    return facility;
+  factory Asset.fromJson(Map<String, dynamic> json) {
```
This overworks the asset class. Rather create a JsonAssetConverter class?

mmathis21 replied:
Not sure what you mean by overworks. I'm pretty sure this is the standard way of implementation.

vdraganic21 commented on Android/indoor_localization/lib/domain/entities/asset.dart:
```diff
@@ -1,66 +1,101 @@
-  bool isActive;
-  List<int> currentZonesIds;
+  final int id;
+  final String name;
```
Should these attributes be final? Because everything except the id can be changed once the asset updates.

vdraganic21 commented on Android/indoor_localization/lib/domain/entities/asset.dart:
```diff
@@ -1,66 +1,101 @@
-  bool isActiveStatus() {
-    updateData();
-    return isActive;
+  Future<String> getFacilityName() async {
```
This method is redundant. We can get the facility name by calling asset.getCurrentFacility().name.

vdraganic21 commented on Android/indoor_localization/lib/domain/entities/asset.dart:
```diff
@@ -1,66 +1,101 @@
+
+  Future<DateTime> getLastSync() async {
+    final updatedAsset = await updateData();
+    return updatedAsset.lastSync;
```
can just be
await updateData();
return lastSync;

vdraganic21 commented on Android/indoor_localization/lib/domain/entities/asset.dart:
```diff
@@ -1,66 +1,101 @@
+    if (updatedData == null) {
+      return this;
+    }
+    return updatedData;
```
When we already update the data, update all of it. The method should apply the changes.

vdraganic21 commented on Android/indoor_localization/lib/domain/entities/facility.dart:
Make similar changes like in the comments for the Asset class.

vdraganic21 commented on Android/indoor_localization/lib/presentation/assets_page/asset_list_item.dart:
```diff
@@ -32,10 +88,11 @@ class AssetListItem extends StatelessWidget {
             padding: const EdgeInsets.symmetric(horizontal: 8.0),
             child: Text(
-              asset.getCurrentFacility()!.name,
+              facilityName,
```
Why not use asset.getCurrentFacility()!.name?

mmathis21 replied:
getCurrentFacility() returns a facility, not a string and the name getter isn't defined.

vdraganic21 commented on Android/indoor_localization/lib/presentation/facilities_page/facilities_card.dart:
```diff
@@ -65,4 +50,39 @@ class FacilityCard extends StatelessWidget {
+  Widget _buildImage() {
+    try {
+
+      final base64String = facility.imageBase64.replaceFirst(RegExp(r'data:image/[^;]+;base64,'), '');
```
Do the prefix removal in the repository class, so the facility object has a "clean" base64 image from the beginning.

vdraganic21 reviewed (changes requested):
Nice job with starting work on the rest service classes.
I added comments on parts which might be problematic in the future or hard to maintain.
Feel free to write responses. We can also further discuss these points during the next daily meeting.
Keep up the good work 🙌

mmathis21 committed:
Removed comment

mmathis21 committed:
Added fetchFacilityById method.

mmathis21 committed:
Refactored attributes - not final now.
updateData() actually updates the data.
getPosition() now returns offsets, for later use when drawing.
Refactored getLastSync method.

mmathis21 committed:
Added prefix removal to fromJson method.

mmathis21 committed:
_buildImage method uses the cleaned base64.

mmathis21 committed:
Refactored api file to be inside of config foler

mmathis21 committed:
Merge remote-tracking branch 'origin/androidapp-dev' into feature/data-layer-integration

# Conflicts:
#	Android/indoor_localization/lib/main.dart
#	Android/indoor_localization/lib/presentation/reports_page/report.dart
#	Android/indoor_localization/pubspec.yaml

mmathis21 committed:
Fixed conflicts
